pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go

```diff
@@ -531,28 +567,107 @@ func patchHTTPFilters(patchContext networking.EnvoyFilter_PatchContext,
 	}
 }
 
+// orderDependencies recursively orders the patches based on insert before/after positions.
+func orderDependencies(patch *model.EnvoyFilterConfigPatchWrapper, original []*model.EnvoyFilterConfigPatchWrapper,
+	ordered *[]*model.EnvoyFilterConfigPatchWrapper) {
+	for _, lp := range original {
+		if lp == patch {
+			continue
+		}
+		if (patch.Operation == networking.EnvoyFilter_Patch_INSERT_BEFORE && hasHTTPFilterMatch(patch)) ||
+			(patch.Operation == networking.EnvoyFilter_Patch_INSERT_AFTER && hasHTTPFilterMatch(patch)) {
+			filter := lp.Value.(*http_conn.HttpFilter)
+			match := patch.Match.GetListener().FilterChain.Filter.SubFilter
+			if match != nil && nameMatches(match.Name, filter.Name) {
+				orderDependencies(lp, original, ordered)
+			}
+		}
+	}
+	if !contains(*ordered, patch) {
+		*ordered = append(*ordered, patch)
+	}
+}
+
+func contains(patches []*model.EnvoyFilterConfigPatchWrapper, patch *model.EnvoyFilterConfigPatchWrapper) bool {
+	for _, p := range patches {
+		if p == patch {
+			return true
+		}
+	}
+	return false
+}
+
+// hasCyclicDependencies checks if the patches have cyclic dependencies.
+// Inspired by https://github.com/dnaeon/go-dependency-graph-algorithm/blob/master/dependency-graph.go#L56.
+func hasCyclicDependencies(original []*model.EnvoyFilterConfigPatchWrapper) bool {
+	dependencies := make(map[string]sets.Set)
+	// First collect all dependencies for every patch.
+	for _, lp := range original {
+		lfilter := lp.Value.(*http_conn.HttpFilter)
+		deps := sets.NewSet()
+		for _, rp := range original {
+			if lp == rp {
+				continue
+			}
+			rmatch := rp.Match.GetListener().FilterChain.Filter.SubFilter
+			if rmatch != nil && nameMatches(rmatch.Name, lfilter.Name) {
+				deps.Insert(rp.Value.(*http_conn.HttpFilter).Name)
+			}
+		}
+		dependencies[lfilter.Name] = deps
+	}
+
+	// Iteratively find and remove patches which have no dependencies. If at some point
+	// there are still patches in the dependencies and we cannot find patches without
+	// dependencies, that means we have a circular dependency.
+	for len(dependencies) != 0 {
+		// Get all filters from the dependencies which have no dependencies
+		independent := sets.NewSet()
```

Comment:
I think we already have converted partial order based on dependencies to full order here.
if we just append “independent” in a top level array, that array gives us the order in which filters should be applied. Every item in the list is independent of items that come later.

We do not need the 2nd orderDependencies function.

Reply:
Thinking about it more, the children depend on the parent. So the parent must be evaluated first.

Len(deps)==0 means that no other filter depends on **it**
that does not mean the filter itself is independent. It could depend on something else right?

I think if we can build a tree after we confirm that there are no cycles, then just breadth first traversal is the order we want.

Reply:
Yes. We can not use that map as is as but we can build a tree after there are no cycles and follow breath first travels. Let me try that.